pkg/controller/certificatesigningrequest/approver_controller.go

```diff
@@ -0,0 +1,335 @@
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Received unexpected object: %v", obj)
+			return
+		}
+		csr, ok = deletedState.Obj.(*certificatesv1.CertificateSigningRequest)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-CertificateSigningRequest object: %v", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(csr.Name)
+}
+
+func (c *CSRApproverController) syncCSR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).Infof("Finished syncing CertificateSigningRequest %s. (%v)", key, d)
+	}()
+
+	csr, err := c.csrLister.Get(key)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	for _, a := range c.approvers {
+		if a.recognize(csr) {
+			return a.verify(csr)
+		}
+	}
+	klog.Warningf("unrecognized CertificateSigningRequest %s", key)
+	return nil
+}
+
+func appendApprovalCondition(csr *certificatesv1.CertificateSigningRequest, message string) {
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:    certificatesv1.CertificateApproved,
+		Status:  corev1.ConditionTrue,
+		Reason:  "AutoApproved",
+		Message: message,
+	})
+}
+
+func appendDeniedCondition(csr *certificatesv1.CertificateSigningRequest, message string) {
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:    certificatesv1.CertificateDenied,
+		Status:  corev1.ConditionTrue,
+		Reason:  "ValidationFailed",
+		Message: message,
+	})
+}
+
+func decodeCertificateRequest(pemBytes []byte) (*x509.CertificateRequest, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != certutil.CertificateRequestBlockType {
+		err := fmt.Errorf("PEM block type must be %s", certutil.CertificateRequestBlockType)
+		return nil, err
+	}
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+func (c *CSRApproverController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.syncCSR(key.(string))
+	if err != nil {
+		// Put the item back in the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync CertificateSigningRequest", "CertificateSigningRequest", key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+type ipsecCertificateApprover struct {
+	client        clientset.Interface
+	approvePolicy config.AutoApprovePolicy
+}
+
+var _ approver = (*ipsecCertificateApprover)(nil)
+
+func (ic *ipsecCertificateApprover) recognize(csr *certificatesv1.CertificateSigningRequest) bool {
+	if csr.Spec.SignerName != antreaSignerName {
+		return false
+	}
+	for _, usage := range csr.Spec.Usages {
+		if !ipsecTunnelUsages.Has(string(usage)) {
```

Comment:
this is redundant with the check in verify? I think it could use signer for recognization and verify for usage checking.